### ISSUE TITLE
Deep Q Controller Possible Error

### ISSUE THREAD

**Issue by Mazecreator** [nivwusquorum/tensorflow-deepq#27]
Thanks for the head start on RL with your DeepQ work.  I am relatively new with RL and I was trying to get a system to converge for the longest time using your DeepQ controller, but it kept tending to 0 in total reward.  My environment gives positive and negative reward, but almost always "converged" to 0 total reward (lowest energy?).

After re-reviewing many RL examples and TensorFlow, I think I found the issue which was surprising to say the least.  I think it is related to TensorFlow's automatic calculation of the gradients.  I feel the error is in this line:

> **temp_diff = self.value_given_action - self.future_reward**

I my mind this is the difference of (Y(x) - Yexpected).  The derivative of this is ultimately +Y(x)/dCost and I think this forces the solution "away" from the minimum.  This in turn increases cost and forced my system to eventually decide to take no action at all (best convergence case it could find).  So I reversed the terms according to most of the literature to (Yexpected - Y(x)) and sure enough, reward would grow positive and converge.

This may not affect examples where the data is all positive and you possible stop early enough with a slow learning rate.  So changing this line to read as follows may improve the algorithm.  This might also fix you Continuous solution as well if that had +/- rewards:

> **temp_diff = self.future_reward - self.value_given_action**

**Comment by siemanko** [comment]
Thanks for investigating this!

Because of the fast the temp_diff is squared later this should not matter, i.e. the gradient ends up being the same for `self.value_given_action` regardless of where it is in tempdiff:

```
e = (a-b)^2
de/da = 2(a-b) * d(a-b)/da = 2 (a-b)
```

or if you swap

```
e = (b-a)^2
de/da = 2(b-a) * d(b-a)/da = 2(a-b) * -1 = 2(a-b)
```

so the gradients come out being the same. Unless TF is doing something crazy. But I am reluctant to believe that TF would do something that fundamentally wrong.

Can you provide a reproducible experiment where one version consistently outperforms the other for a couple of different random seeds?

**Comment by Mazecreator** [comment]
Hummm... Sorry, I switched it back and cannot reproduce the problem.  My model must have been underfit originally and I must have modified it around the same time I tried changing the sign.  I was surprised as well, but I figured it may have something to do with the Stop Gradient command causing the error indirectly.
